### DISCUSSION TITLE
Detailed Mapping between GCP Structured Log and OTEL Standard ?

### DISCUSSION THREAD

**Discussion by rober15** [open-telemetry/community#1466] [Community Discussions]
Where can i find the detailed mapping logic between GCP Structured Log Standard (priopriatary):
https://cloud.google.com/logging/docs/structured-logging
and:
https://opentelemetry.io/docs/reference/specification/logs/data-model/  ?

I found this info but that is to high level:
https://github.com/open-telemetry/oteps/blob/main/text/logs/0097-log-data-model.md#google-cloud-logging

What is too high level ?

- No details how the different resource names in gcp map into the resource key values in OTEL ?
https://cloud.google.com/monitoring/api/resources

- Unclear how in OTEL with just Body it will differ between the different types of bodies in GCP:
jsonPayload => holds a json (structured log)
testPayload => holds just a string

- Unclear how in detail labels get mapped ?
Labels go into Attributes but how ?

- TraceId look like mapped 1:1 from trace into traceid but that is not correct as google includes more then just traceId in the field: trace. Gcp Trace Example: "trace": "projects/my-projectid/traces/06796866738c859f2f19b7cfb3214824",
but what should be mapped to traceId is just: 06796866738c859f2f19b7cfb3214824

....

**Comment by trask** [comment]
cc @jsuereth @dashpole

**Comment by dashpole** [comment]
@damemi does this exist somewhere?

**Reply by damemi** [reply]
No, but it would be good to have. I don't think we document our monitored resource mapping for any signals. For the others, we could probably just add some more detail to the data model:

* For json/text payload, that is just based on the [value type of the log body from otlp](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/2050f26477c2e8f7a90f901c49e4ac9a40a99eb4/exporter/collector/logs.go#L521-L527).
* labels are string representations of attributes, except the few special attributes with a `gcp.*` prefix that map to a GCP field
* traceId is the trace ID value with `projects/$project_id/traces/` [prepended](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/2050f26477c2e8f7a90f901c49e4ac9a40a99eb4/exporter/collector/logs.go#L422)

**Reply by rober15** [reply]
Many thanks for your reply.
It would be great we get more details about mapping as without that it is really hard to understand what happens and we should get this info as part of documentation instead of scanning several code places to figure out.

I looked now [value type of the log body from otlp](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/2050f26477c2e8f7a90f901c49e4ac9a40a99eb4/exporter/collector/logs.go#L521-L527) but still do not understand how we can influence in the logging itself that it to be one of the 3 types.

It looks the mapping is:
- Binary => proto_payload
- String => text_payload
- Map => json_payload

So  how to log that the proper logBody.Type() is set to perform the right mapping ?
Many thanks for your help.

**Reply by trask** [reply]
hi @rober15, i'd suggest asking in https://github.com/GoogleCloudPlatform/opentelemetry-operations-go to get the best answer
